_posts/2020-05-06-instrumenting_phoenix_with_telemetry_part_three.md

```diff
@@ -0,0 +1,469 @@
+
+For example:
+
+```elixir
+# lib/quantum/telemetry.ex
+
+def metrics do
+  [
+    counter(
+      "phoenix.socket_connected.count",
+      tags: [:endpoint]
+    )
+  ]
+end
+```
+
+Now we will increment a StatsD metric every time the socket is joined.
+
+### `Phoenix.Channel` Telemetry Events
+
+The `Phoenix.Channel.Server` module executes two Telemetry events--one when the channel is joined and one whenever the channel invokes `handle_info/2`.
+
+We can see the `[:phoenix, :channel_joined]` Telemetry event in source code [here](https://github.com/phoenixframework/phoenix/blob/8a4aa4eed0de69f94ab09eca157c87d9bd204168/lib/phoenix/channel/server.ex#L302):
+
+```elixir
+# phoenix/lib/phoenix/channel/server.ex
+
+def handle_info({:join, __MODULE__}, {auth_payload, {pid, _} = from, socket}) do
+  %{channel: channel, topic: topic, private: private} = socket
+
+  start = System.monotonic_time()
+  {reply, state} = channel_join(channel, topic, auth_payload, socket)
+  duration = System.monotonic_time() - start
+  metadata = %{params: auth_payload, socket: socket, result: elem(reply, 0)}
+  :telemetry.execute([:phoenix, :channel_joined], %{duration: duration}, metadata)
+  GenServer.reply(from, reply)
+  state
+end
+```
+
+And we can see the `[:phoenix, channel_handled_in]` event [here](https://github.com/phoenixframework/phoenix/blob/8a4aa4eed0de69f94ab09eca157c87d9bd204168/lib/phoenix/channel/server.ex#L319)
+
+```elixir
+# phoenix/lib/phoenix/channel/server.ex
+
+def handle_info(
+    %Message{topic: topic, event: event, payload: payload, ref: ref},
+    %{topic: topic} = socket
+  ) do
+  start = System.monotonic_time()
+
+  result = socket.channel.handle_in(event, payload, put_in(socket.ref, ref))
+  duration = System.monotonic_time() - start
+  metadata = %{ref: ref, event: event, params: payload, socket: socket}
+
+  :telemetry.execute([:phoenix, :channel_handled_in], %{duration: duration}, metadata)
+
+  handle_in(result)
+end
+```
+
+We can add some metrics reporting for these events by defining metrics in `Quantum.Telemetry` for either of the `"phoenix.channel_joined"` and `"phoenix.channel_handled_in"` events.
+
+Now that we've taken a brief tour of Phoenix Telemetry events, let's hook up some reporting for Ecto events.
+
+## Ecto Telemetry Events
+
+Ecto provides some out-of-the-box instrumentation for queries. Let's take a look at and define metrics for some of these Telemetry events now.
+
+Ecto will execute a Telemetry event, [`[:my_app, :repo, :query]`](https://github.com/elixir-ecto/ecto/blob/2aca7b28eef486188be66592055c7336a80befe9/lib/ecto/repo.ex#L117) for every query sent to the Ecto adapter. It will emit this event with a measurement map and a metadata map.
```

Comment:
```suggestion
Ecto will execute a Telemetry event, [`[:my_app, :repo, :query]`](https://github.com/elixir-ecto/ecto/blob/2aca7b28eef486188be66592055c7336a80befe9/lib/ecto/repo.ex#L120) for every query sent to the Ecto adapter. It will emit this event with a measurement map and a metadata map.
```

The line number changed 😬